signer/src/transaction_coordinator.rs

```diff
@@ -485,43 +487,69 @@ where
         coordinator_state_machine: &mut wsts_state_machine::CoordinatorStateMachine,
         txid: bitcoin::Txid,
     ) -> Result<wsts::taproot::SchnorrProof, Error> {
+        let mut signal_rx = self.context.get_signal_receiver();
+
+        // We'll poll both the network and the signal channel for messages
+        // from our own tx signer. We'll store any received messages here.
+        let mut signer_messages: Vec<Signed<SignerMessage>> = vec![];
+
         loop {
-            let msg = self.network.receive().await?;
+            // Empty the signal channel and collect all messages generated by
+            // our own transaction signer.
+            while let Ok(msg) = signal_rx.try_recv() {
+                if let SignerSignal::Event(SignerEvent::TxSigner(
+                    TxSignerEvent::MessageGenerated(msg),
+                )) = msg
+                {
+                    signer_messages.push(msg);
+                }
+            }
 
-            if &msg.bitcoin_chain_tip != bitcoin_chain_tip {
-                tracing::warn!(?msg, "concurrent wsts signing round message observed");
-                continue;
+            // Check the network for new messages. We don't have a `try_receive()`
+            // equivilent for the network, so we use a timeout to avoid blocking.
+            if let Ok(msg) =
+                tokio::time::timeout(Duration::from_millis(10), self.network.receive()).await
+            {
+                signer_messages.push(msg?);
             }
```

Comment:
This could have been done more concisely with a `tokio::select!` on `signal_rx.recv()` and `network.receive()` and skipped the buffer/for-loop, but I opted not to due to the number of messages flying around and the context switching it would entail.

Reply:
Actually, I need to revisit this part of the codebase anyway so I'll give this a quick look over to see if it is indeed better.

Reply:
It's a lot cleaner looking, I just did it. But then the `process_complete_deposit()` integration test started failing.

What I think might be happening is that `network.receive()` and `signal_rx.recv()` start processing concurrently in the test and one of them gets aborted (by `select!`), meaning the coordinator then misses the message from the receive that got aborted.

I didn't want to spend time on it pre-demo, so I'll go ahead and merge this solution for now and we can revisit :+1: